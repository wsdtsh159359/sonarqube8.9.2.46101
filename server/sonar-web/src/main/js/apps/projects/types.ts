/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../types/component';

export interface Project {
  analysisDate?: string;
  isFavorite?: boolean;
  key: string;
  leakPeriodDate?: string;
  measures: T.Dict<string>;
  name: string;
  projects?: number;
  qualifier: ComponentQualifier;
  tags: string[];
  visibility: T.Visibility;
  needIssueSync?: boolean;
}

export interface Facet {
  [value: string]: number;
}

export interface Facets {
  [property: string]: Facet;
}
