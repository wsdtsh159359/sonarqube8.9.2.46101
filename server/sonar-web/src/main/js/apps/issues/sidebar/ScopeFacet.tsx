/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { without } from 'lodash';
import * as React from 'react';
import QualifierIcon from 'sonar-ui-common/components/icons/QualifierIcon';
import { translate } from 'sonar-ui-common/helpers/l10n';
import FacetBox from '../../../components/facet/FacetBox';
import FacetHeader from '../../../components/facet/FacetHeader';
import FacetItem from '../../../components/facet/FacetItem';
import FacetItemsList from '../../../components/facet/FacetItemsList';
import MultipleSelectionHint from '../../../components/facet/MultipleSelectionHint';
import { SOURCE_SCOPES } from '../../../helpers/constants';
import { formatFacetStat, Query } from '../utils';

export interface ScopeFacetProps {
  fetching: boolean;
  onChange: (changes: Partial<Query>) => void;
  onToggle: (property: string) => void;
  open: boolean;
  scopes: string[];
  stats: T.Dict<number> | undefined;
}

export default function ScopeFacet(props: ScopeFacetProps) {
  const { fetching, open, scopes = [], stats = {} } = props;
  const values = scopes.map(scope => translate('issue.scope', scope));

  return (
    <FacetBox property="scopes">
      <FacetHeader
        fetching={fetching}
        name={translate('issues.facet.scopes')}
        onClear={() => props.onChange({ scopes: [] })}
        onClick={() => props.onToggle('scopes')}
        open={open}
        values={values}
      />

      {open && (
        <>
          <FacetItemsList>
            {SOURCE_SCOPES.map(({ scope, qualifier }) => {
              const active = scopes.includes(scope);
              const stat = stats[scope];

              return (
                <FacetItem
                  active={active}
                  key={scope}
                  name={
                    <span className="display-flex-center">
                      <QualifierIcon className="little-spacer-right" qualifier={qualifier} />{' '}
                      {translate('issue.scope', scope)}
                    </span>
                  }
                  onClick={(itemValue: string, multiple: boolean) => {
                    if (multiple) {
                      props.onChange({
                        scopes: active ? without(scopes, itemValue) : [...scopes, itemValue]
                      });
                    } else {
                      props.onChange({
                        scopes: active && scopes.length === 1 ? [] : [itemValue]
                      });
                    }
                  }}
                  stat={formatFacetStat(stat)}
                  value={scope}
                />
              );
            })}
          </FacetItemsList>
          <MultipleSelectionHint options={Object.keys(stats).length} values={scopes.length} />
        </>
      )}
    </FacetBox>
  );
}
