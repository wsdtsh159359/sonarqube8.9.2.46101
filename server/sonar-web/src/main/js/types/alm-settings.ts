/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const enum AlmKeys {
  Azure = 'azure',
  BitbucketServer = 'bitbucket',
  BitbucketCloud = 'bitbucketcloud',
  GitHub = 'github',
  GitLab = 'gitlab'
}

export interface AlmBindingDefinition {
  key: string;
  url?: string;
}

export interface AzureBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface BitbucketBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url: string;
}

export interface BitbucketCloudBindingDefinition extends AlmBindingDefinition {
  clientId: string;
  clientSecret: string;
  workspace: string;
}

export interface GithubBindingDefinition extends AlmBindingDefinition {
  appId: string;
  clientId: string;
  clientSecret: string;
  privateKey: string;
  url: string;
}

export interface GitlabBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface ProjectAlmBindingResponse {
  alm: AlmKeys;
  key: string;
  repository?: string;
  slug?: string;
  summaryCommentEnabled?: boolean;
  monorepo: boolean;
}

export interface ProjectAzureBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.Azure;
  repository: string;
  slug: string;
  url: string;
}

export interface ProjectBitbucketBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.BitbucketServer;
  repository: string;
  slug: string;
}

export interface ProjectGitHubBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.GitHub;
  repository: string;
}

export interface ProjectGitLabBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.GitLab;
  repository: string;
  url: string;
}

export interface ProjectAlmBindingParams {
  almSetting: string;
  project: string;
  monorepo: boolean;
}

export interface AzureProjectAlmBindingParams extends ProjectAlmBindingParams {
  projectName: string;
  repositoryName: string;
}

export interface BitbucketProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  slug: string;
}

export interface BitbucketCloudProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
}

export interface GithubProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  summaryCommentEnabled: boolean;
}

export interface GitlabProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository?: string;
}

export interface AlmSettingsInstance {
  alm: AlmKeys;
  key: string;
  url?: string;
}

export interface AlmSettingsBindingDefinitions {
  [AlmKeys.Azure]: AzureBindingDefinition[];
  [AlmKeys.BitbucketServer]: BitbucketBindingDefinition[];
  [AlmKeys.BitbucketCloud]: BitbucketCloudBindingDefinition[];
  [AlmKeys.GitHub]: GithubBindingDefinition[];
  [AlmKeys.GitLab]: GitlabBindingDefinition[];
}

export interface AlmSettingsBindingStatus {
  alertSuccess: boolean;
  failureMessage: string;
  type: AlmSettingsBindingStatusType;
}

export enum AlmSettingsBindingStatusType {
  Validating,
  Success,
  Failure,
  Warning
}

export function isProjectBitbucketBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectBitbucketBindingResponse {
  return binding.alm === AlmKeys.BitbucketServer;
}

export function isProjectGitHubBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitHubBindingResponse {
  return binding.alm === AlmKeys.GitHub;
}

export function isProjectGitLabBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitLabBindingResponse {
  return binding.alm === AlmKeys.GitLab;
}

export function isProjectAzureBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectAzureBindingResponse {
  return binding.alm === AlmKeys.Azure;
}

export function isBitbucketBindingDefinition(
  binding?: AlmBindingDefinition & { url?: string }
): binding is BitbucketBindingDefinition {
  return binding !== undefined && binding.url !== undefined;
}

export function isBitbucketCloudBindingDefinition(
  binding?: AlmBindingDefinition & { clientId?: string; workspace?: string }
): binding is BitbucketCloudBindingDefinition {
  return binding !== undefined && binding.clientId !== undefined && binding.workspace !== undefined;
}

export function isGithubBindingDefinition(
  binding?: AlmBindingDefinition & { appId?: string; url?: string }
): binding is GithubBindingDefinition {
  return binding !== undefined && binding.appId !== undefined && binding.url !== undefined;
}

export function isGitLabBindingDefinition(
  binding?: AlmBindingDefinition | GithubBindingDefinition | BitbucketCloudBindingDefinition
): binding is GitlabBindingDefinition {
  // There's too much overlap with the others. We must not only validate that certain fields are
  // present, we must also validate that others are NOT present. And even so, we cannot be 100%
  // sure, as right now, Azure, Bitbucket Server, and GitLab have the same signature.
  return (
    binding !== undefined &&
    (binding as GithubBindingDefinition).appId === undefined &&
    (binding as BitbucketCloudBindingDefinition).workspace === undefined
  );
}
