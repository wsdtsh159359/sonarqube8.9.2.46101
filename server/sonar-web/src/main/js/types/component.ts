/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export enum Visibility {
  Public = 'public',
  Private = 'private'
}

export enum ComponentQualifier {
  Application = 'APP',
  Directory = 'DIR',
  Developper = 'DEV',
  File = 'FIL',
  Portfolio = 'VW',
  Project = 'TRK',
  SubPortfolio = 'SVW',
  SubProject = 'BRC',
  TestFile = 'UTS'
}

export enum ProjectKeyValidationResult {
  Valid = 'valid',
  Empty = 'empty',
  TooLong = 'too_long',
  InvalidChar = 'invalid_char',
  OnlyDigits = 'only_digits'
}

export interface TreeComponent extends T.LightComponent {
  id?: string;
  name: string;
  path?: string;
  refId?: string;
  refKey?: string;
  tags?: string[];
  visibility: T.Visibility;
}

export interface TreeComponentWithPath extends TreeComponent {
  path: string;
}

export function isPortfolioLike(
  componentQualifier?: string | ComponentQualifier
): componentQualifier is ComponentQualifier.Portfolio | ComponentQualifier.SubPortfolio {
  return Boolean(
    componentQualifier &&
      [
        ComponentQualifier.Portfolio.toString(),
        ComponentQualifier.SubPortfolio.toString()
      ].includes(componentQualifier)
  );
}

export function isApplication(
  componentQualifier?: string | ComponentQualifier
): componentQualifier is ComponentQualifier.Application {
  return componentQualifier === ComponentQualifier.Application;
}
