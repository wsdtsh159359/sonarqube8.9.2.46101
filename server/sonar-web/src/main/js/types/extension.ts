/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { InjectedIntl } from 'react-intl';
import { Store as ReduxStore } from 'redux';
import { Theme } from 'sonar-ui-common/components/theme';
import { Location, Router } from '../components/hoc/withRouter';
import { Store } from '../store/rootReducer';
import { L10nBundle } from './l10n';

export interface ExtensionStartMethod {
  (params: ExtensionStartMethodParameter | string): ExtensionStartMethodReturnType;
}

export interface ExtensionStartMethodParameter {
  store: ReduxStore<Store, any>;
  el: HTMLElement | undefined | null;
  currentUser: T.CurrentUser;
  intl: InjectedIntl;
  location: Location;
  router: Router;
  theme: Theme;
  baseUrl: string;
  l10nBundle: L10nBundle;
}

export type ExtensionStartMethodReturnType = React.ReactNode | Function | void | undefined | null;
