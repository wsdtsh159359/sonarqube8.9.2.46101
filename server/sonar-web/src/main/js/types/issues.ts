/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export enum IssueType {
  CodeSmell = 'CODE_SMELL',
  Vulnerability = 'VULNERABILITY',
  Bug = 'BUG',
  SecurityHotspot = 'SECURITY_HOTSPOT'
}

export enum IssueScope {
  Main = 'MAIN',
  Test = 'TEST'
}

interface Comment {
  createdAt: string;
  htmlText: string;
  key: string;
  login: string;
  markdown: string;
  updatable: boolean;
}

export interface RawIssue {
  assignee?: string;
  author?: string;
  comments?: Array<Comment>;
  component: string;
  flows?: Array<{
    // `componentName` is not available in RawIssue
    locations?: Array<T.Omit<T.FlowLocation, 'componentName'>>;
  }>;
  key: string;
  line?: number;
  project: string;
  rule: string;
  severity: string;
  status: string;
  subProject?: string;
  textRange?: T.TextRange;
}

export interface IssueResponse {
  components?: Array<{ key: string; name: string }>;
  issue: RawIssue;
  rules?: Array<{}>;
  users?: Array<T.UserBase>;
}

export interface RawIssuesResponse {
  components: ReferencedComponent[];
  effortTotal: number;
  facets: RawFacet[];
  issues: RawIssue[];
  languages: ReferencedLanguage[];
  paging: T.Paging;
  rules?: Array<{}>;
  users?: Array<T.UserBase>;
}

export interface FetchIssuesPromise {
  components: ReferencedComponent[];
  effortTotal: number;
  facets: RawFacet[];
  issues: T.Issue[];
  languages: ReferencedLanguage[];
  paging: T.Paging;
  rules: ReferencedRule[];
  users: T.UserBase[];
}

export interface ReferencedComponent {
  key: string;
  name: string;
  path?: string;
  uuid: string;
}

export interface ReferencedLanguage {
  name: string;
}

export interface ReferencedRule {
  langName?: string;
  name: string;
}

export interface RawFacet {
  property: string;
  values: Array<{ val: string; count: number }>;
}

export interface Facet {
  [value: string]: number;
}
