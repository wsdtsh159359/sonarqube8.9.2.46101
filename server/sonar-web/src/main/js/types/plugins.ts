/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export interface Plugin {
  key: string;
  name: string;
  category?: string;
  description?: string;
  editionBundled?: boolean;
  license?: string;
  organizationName?: string;
  homepageUrl?: string;
  organizationUrl?: string;
  issueTrackerUrl?: string;
  termsAndConditionsUrl?: string;
}

export interface PendingPluginResult {
  installing: PendingPlugin[];
  updating: PendingPlugin[];
  removing: PendingPlugin[];
}

export interface AvailablePlugin extends Plugin {
  release: Release;
  update: Update;
}

export interface PendingPlugin extends Plugin {
  version: string;
  implementationBuild: string;
}

export interface InstalledPlugin extends PendingPlugin {
  documentationPath?: string;
  issueTrackerUrl?: string;
  filename: string;
  hash: string;
  sonarLintSupported: boolean;
  updatedAt: number;
  updates?: Update[];
}

export interface Release {
  version: string;
  date: string;
  description?: string;
  changeLogUrl?: string;
}

export interface Update {
  status: string;
  release?: Release;
  requires: Plugin[];
  previousUpdates?: Update[];
}

export enum PluginType {
  Bundled = 'BUNDLED',
  External = 'EXTERNAL'
}

export enum RiskConsent {
  Accepted = 'ACCEPTED',
  NotAccepted = 'NOT_ACCEPTED',
  Required = 'REQUIRED'
}

export function isAvailablePlugin(plugin: Plugin): plugin is AvailablePlugin {
  return (plugin as any).release !== undefined;
}

export function isInstalledPlugin(plugin: Plugin): plugin is InstalledPlugin {
  return isPendingPlugin(plugin) && (plugin as any).updatedAt !== undefined;
}

export function isPendingPlugin(plugin: Plugin): plugin is PendingPlugin {
  return (plugin as any).version !== undefined;
}
