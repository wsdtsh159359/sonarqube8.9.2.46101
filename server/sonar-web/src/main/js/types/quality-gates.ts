/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { BranchLike } from './branch-like';

export interface QualityGateProjectStatus {
  conditions?: QualityGateProjectStatusCondition[];
  ignoredConditions: boolean;
  status: T.Status;
}

export interface QualityGateProjectStatusCondition {
  actualValue: string;
  comparator: string;
  errorThreshold: string;
  metricKey: string;
  periodIndex: number;
  status: T.Status;
}

export interface QualityGateApplicationStatus {
  metrics: T.Metric[];
  projects: QualityGateApplicationStatusChildProject[];
  status: T.Status;
}

export interface QualityGateApplicationStatusCondition {
  comparator: string;
  errorThreshold?: string;
  metric: string;
  periodIndex?: number;
  onLeak?: boolean;
  status: T.Status;
  value: string;
  warningThreshold?: string;
}

export interface QualityGateApplicationStatusChildProject {
  conditions: QualityGateApplicationStatusCondition[];
  key: string;
  name: string;
  status: T.Status;
}

export interface QualityGateStatus {
  failedConditions: QualityGateStatusConditionEnhanced[];
  ignoredConditions?: boolean;
  key: string;
  name: string;
  status: T.Status;
  branchLike?: BranchLike;
}

export interface QualityGateStatusCondition {
  actual?: string;
  error?: string;
  level: T.Status;
  metric: string;
  op: string;
  period?: number;
  warning?: string;
}

export interface QualityGateStatusConditionEnhanced extends QualityGateStatusCondition {
  measure: T.MeasureEnhanced;
}
