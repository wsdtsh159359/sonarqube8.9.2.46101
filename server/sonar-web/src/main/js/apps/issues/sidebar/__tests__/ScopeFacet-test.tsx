/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import FacetHeader from '../../../../components/facet/FacetHeader';
import FacetItem from '../../../../components/facet/FacetItem';
import { IssueScope } from '../../../../types/issues';
import ScopeFacet, { ScopeFacetProps } from '../ScopeFacet';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ open: true })).toMatchSnapshot('open');
  expect(shallowRender({ open: true, scopes: [IssueScope.Main] })).toMatchSnapshot('active facet');
  expect(shallowRender({ open: true, stats: { [IssueScope.Main]: 0 } })).toMatchSnapshot(
    'disabled facet'
  );
});

it('should correctly handle facet header clicks', () => {
  const onChange = jest.fn();
  const onToggle = jest.fn();
  const wrapper = shallowRender({ onChange, onToggle });

  wrapper.find(FacetHeader).props().onClear!();
  expect(onChange).toBeCalledWith({ scopes: [] });

  wrapper.find(FacetHeader).props().onClick!();
  expect(onToggle).toBeCalledWith('scopes');
});

it('should correctly handle facet item clicks', () => {
  const wrapper = shallowRender({ open: true, scopes: [IssueScope.Main] });
  const onChange = jest.fn(({ scopes }) => wrapper.setProps({ scopes }));
  wrapper.setProps({ onChange });

  clickFacetItem(wrapper, IssueScope.Test);
  expect(onChange).toHaveBeenLastCalledWith({ scopes: [IssueScope.Test] });

  clickFacetItem(wrapper, IssueScope.Test);
  expect(onChange).toHaveBeenLastCalledWith({ scopes: [] });

  clickFacetItem(wrapper, IssueScope.Test, true);
  clickFacetItem(wrapper, IssueScope.Main, true);
  expect(onChange).toHaveBeenLastCalledWith({
    scopes: expect.arrayContaining([IssueScope.Main, IssueScope.Test])
  });

  clickFacetItem(wrapper, IssueScope.Test, true);
  expect(onChange).toHaveBeenLastCalledWith({ scopes: [IssueScope.Main] });
});

function clickFacetItem(
  wrapper: ShallowWrapper<ScopeFacetProps>,
  scope: IssueScope,
  multiple = false
) {
  return wrapper
    .find(FacetItem)
    .filterWhere(f => f.key() === scope)
    .props()
    .onClick(scope, multiple);
}

function shallowRender(props: Partial<ScopeFacetProps> = {}) {
  return shallow<ScopeFacetProps>(
    <ScopeFacet
      fetching={true}
      onChange={jest.fn()}
      onToggle={jest.fn()}
      open={false}
      scopes={[]}
      stats={{}}
      {...props}
    />
  );
}
