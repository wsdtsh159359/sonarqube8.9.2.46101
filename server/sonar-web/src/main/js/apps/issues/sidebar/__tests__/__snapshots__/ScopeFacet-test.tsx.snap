// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: active facet 1`] = `
<FacetBox
  property="scopes"
>
  <FacetHeader
    fetching={true}
    name="issues.facet.scopes"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={
      Array [
        "issue.scope.MAIN",
      ]
    }
  />
  <FacetItemsList>
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="MAIN"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="FIL"
          />
           
          issue.scope.MAIN
        </span>
      }
      onClick={[Function]}
      value="MAIN"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="TEST"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="UTS"
          />
           
          issue.scope.TEST
        </span>
      }
      onClick={[Function]}
      value="TEST"
    />
  </FacetItemsList>
  <MultipleSelectionHint
    options={0}
    values={1}
  />
</FacetBox>
`;

exports[`should render correctly: default 1`] = `
<FacetBox
  property="scopes"
>
  <FacetHeader
    fetching={true}
    name="issues.facet.scopes"
    onClear={[Function]}
    onClick={[Function]}
    open={false}
    values={Array []}
  />
</FacetBox>
`;

exports[`should render correctly: disabled facet 1`] = `
<FacetBox
  property="scopes"
>
  <FacetHeader
    fetching={true}
    name="issues.facet.scopes"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="MAIN"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="FIL"
          />
           
          issue.scope.MAIN
        </span>
      }
      onClick={[Function]}
      stat={0}
      value="MAIN"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="TEST"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="UTS"
          />
           
          issue.scope.TEST
        </span>
      }
      onClick={[Function]}
      value="TEST"
    />
  </FacetItemsList>
  <MultipleSelectionHint
    options={1}
    values={0}
  />
</FacetBox>
`;

exports[`should render correctly: open 1`] = `
<FacetBox
  property="scopes"
>
  <FacetHeader
    fetching={true}
    name="issues.facet.scopes"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="MAIN"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="FIL"
          />
           
          issue.scope.MAIN
        </span>
      }
      onClick={[Function]}
      value="MAIN"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="TEST"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <QualifierIcon
            className="little-spacer-right"
            qualifier="UTS"
          />
           
          issue.scope.TEST
        </span>
      }
      onClick={[Function]}
      value="TEST"
    />
  </FacetItemsList>
  <MultipleSelectionHint
    options={0}
    values={0}
  />
</FacetBox>
`;
